pyanaconda/modules/boss/module_manager.py

```diff
@@ -19,111 +19,106 @@
-        dbus = DBus.get_dbus_proxy()
-        names = dbus.ListActivatableNames()
-        for name in names:
-            if name.startswith(DBUS_ADDON_NAMESPACE):
-                self._addon_module_services.append(name)
+        if self.check_modules_availability():
+            log.info("All modules are ready now.")
 
+    def _process_module_is_available(self, observer):
+        """Process the service_available signal."""
+        log.debug("%s is available", observer)
+        observer.proxy.EchoString("Boss says hi!")
 
-    def check_no_modules_are_running(self):
-        dbus = DBus.get_dbus_proxy()
-        for service in self.expected_module_services:
-            if dbus.NameHasOwner(service):
-                log.error("service %s has unexpected owner", service)
+    def _process_module_is_unavailable(self, observer):
+        """Process the service_unavailable signal."""
+        log.debug("%s is unavailable", observer)
 
-    def start_modules(self):
-        """Starts anaconda modules (including addons)."""
-        log.debug("starting modules")
-        self.check_no_modules_are_running()
+    def check_modules_availability(self):
+        """Check if all modules are available.
 
-        dbus = DBus.get_dbus_proxy()
-        for service in self.expected_module_services:
-            log.debug("Starting %s", service)
-            try:
-                dbus.StartServiceByName(service, 0, callback=self._finish_start_service_cb, callback_args=(service,))
-            except Exception:  # pylint: disable=broad-except
-                self._failed_module_services.append(service)
-                log.exception("module startup failed")
+        :return True if all modules are available, otherwise False
+        """
+        for observer in self.module_observers:
+            if not observer.is_service_available:
+                return False
 
-        log.debug("started all modules")
+        return True
 
     def stop_modules(self):
-        """Tells all running modules to quit."""
-        log.debug("sending Quit to all modules and addons")
-        for service in self.running_module_services:
-            # FIXME: This is just a temporary solution.
-            module = DBus.get_proxy(service, auto_object_path(service))
-            # TODO: async ?
-            # possible reasons:
-            # - module hanging in Quit, deadlocking shutdown
-            try:
-                module.Quit()
-            except Exception:  # pylint: disable=broad-except
-                log.exception("Quit failed for module: %s", service)
+        """Tell all running modules to quit."""
+        log.debug("Stop modules.")
+        for observer in self.module_observers:
+            if not observer.is_service_available:
+                continue
+
+            # Call asynchronously to avoid problems.
+            observer.proxy.Quit(callback=self._stop_modules_callback,
+                                callback_args=(observer,))
```

Comment:
What about catching exception from the observer here?

Reply:
That's why we check if the service is available before we access the proxy.

Reply:
I'm not sure if the checking is a good solution. It is basically race condition. Before we test availability and do that call could happen anything, and thanks to how multi threading works in python it could be a "long time" pause between these two instructions...

Reply:
Well, I think as this is DBUS this is always a race condition - while the chance is small, it's possible the DBUS service/entity will no longer be there once you try to access it even right after a successful `is_service_available()` call.

Even though I guess it could be to tedious to handle this in individual modules, it might make sense to watch for this in Boss and other core places.

Reply:
If a service is supposed to be available and it is not, then it is a bug - see the comment below.

Reply:
OK, then it should be fine.